[Lee Ian Ee] iP

leeianee opened the pull request:
# iP Pull Request

- Bullet1
- Bullet2

1. Numbered 1
2. Numbered 2

```java
String s = "Fenced code block";
```

- [x] Task List Item 1
- [ ] Task List Item 2

😄 

> Block quote

This is a in-line [Hyperlink](https://github.com/)

**Bold**

Commit:
Add Gradle support

leeianee committed:
Add skeletal functionality to Duke.java

leeianee committed:
Add list functionality to Duke.java

leeianee committed:
Add Task.java

leeianee committed:
Add getDescription() to Task.java

leeianee committed:
Add mark as done functionality to Duke.java

leeianee committed:
Add toString to Task.java

leeianee committed:
Add Todo.java

leeianee committed:
Add Deadline.java

leeianee committed:
Add Event.java

leeianee committed:
Add ToDo, Event, Deadline functionality to Duke.java

leeianee committed:
Implemented Automated Text UI Testing

leeianee committed:
Add DukeException.java

leeianee committed:
Add error handling functionality to Duke.java

leeianee committed:
Add Java Collections implementation

leeianee committed:
Add delete functionality to Duke.java

leeianee committed:
Add save functionality to Duke.java

leeianee committed:
Add dates and times functionality to Duke.java

leeianee committed:
Merge branch 'branch-Level-8'

leeianee committed:
Add read functionality to Duke.java

leeianee committed:
Merge branch 'branch-Level-7'

leeianee committed:
Add Parser.java

leeianee committed:
Add Storage.java

leeianee committed:
Add TaskList.java

leeianee committed:
Add Ui.java

leeianee committed:
Added more OOP to Duke.java

leeianee committed:
Add duke package

leeianee committed:
Merge branch 'add-gradle-support'

leeianee committed:
Add checkstyle plugin to build.gradle

leeianee committed:
Add JUnit tests

leeianee committed:
Add JAR file

leeianee committed:
Add JavaDocs

leeianee committed:
Comply with coding standard

leeianee committed:
Add find functionality to Duke.java

leeianee committed:
Add checkStyle

kevinchangjk commented on src/main/java/duke/Deadline.java:
```diff
@@ -0,0 +1,42 @@
+        super(description);
+        this.by = by;
+        try {
+            LocalDate d1 = LocalDate.parse("2019-12-01");
```
Hmm I'm not really sure but this doesn't seem to belong here? Maybe it was left here by mistake?

kevinchangjk commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,52 @@
+        Parser parser = new Parser(this.tasks);
+        String input = "";
+        Scanner scan = new Scanner(System.in);
+        input = scan.nextLine();
```
I think you could declare and instantiate `input` on the same line here, to make it more concise.

```suggestion
        String input = scan.nextLine();
```

kevinchangjk commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,52 @@
+    }
+
+    public static void main(String[] args) {
+        new Duke("data/duke.txt").run();
```
In this case, the filepath `"data/duke.txt"` is hard-coded right? Perhaps it can be declared earlier on as a constant?

kevinchangjk commented on src/main/java/duke/DukeException.java:
```diff
@@ -0,0 +1,33 @@
+    }
+
+    /**
+     * Constructor for DukeException for a missing task descriptor.
```
Minor nit, but I think there should be an empty line between the descriptor and params sections here? I noticed that you added them for some other comments, so maybe can be more consistent with the javadocs comments.

kevinchangjk commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,99 @@
+        this.list = list;
+    }
+
+    /**
```
I think this javadocs comment is missing a method descriptor?

kevinchangjk commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,99 @@
+                            this.list.findTask(keyword).printTaskList();
+                        }
+                        break;
+                    default: // Default case
```
This is based on what I understand so far from your code.

You have a set of valid inputs declared earlier, and here you use the switch case statement to check which valid command the input matches, but this is only after an if statement basically checking if the first word matches any valid commands.

If so, then logically (assuming your program works as it should) the default case should never occur right? 

If that is so, I think it would make more sense to simply have it be empty, and instead, at **line 91** below (reached after there is an invalid input), the error message can be included as `"Wrong input"`, instead of being blank. Would that be better?

kevinchangjk commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,99 @@
+                        System.out.println("Wrong input");
+                    }
+                } else {
+                    throw new DukeException(); // Invalid input
```
As mentioned above, maybe something like this?

```suggestion
                    throw new DukeException("Wrong input"); // Invalid input
```

kevinchangjk reviewed (commented):
Quite detailed and thorough work. I think this is great as a product, but perhaps the code standards can be a bit more consistent. Good job!

leeianee committed:
Add GUI

Chustinjeng commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,89 @@
+        this.list = list;
+    }
+
+    /**
```
missing method descriptor in javadocs

Chustinjeng commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,141 @@
+     * @param description The description of the Deadline object.
+     * @param by When the Deadline object has to be completed by.
+     */
+    public Task addDeadline(String description, String by) {
+        this.list.add(new Deadline(description, by));
+        return this.list.get(list.size() - 1);
+    }
+
+    /**
+     * Adds an Event object in the TaskList.
+     * @param description The description of the Event object.
+     * @param when When the Event will occur.
+     */
+    public Task addEvent(String description, String when) {
+        this.list.add(new Event(description, when));
+        return this.list.get(list.size() - 1);
+    }
+
```
Can perhaps merge into one addTask method? Both methods have the same parameter types so can maybe use polymorphism

Chustinjeng commented on src/main/java/duke/task/Deadline.java:
```diff
@@ -0,0 +1,50 @@
+        super(description);
+        this.by = by;
+        try {
+            LocalDate d1 = LocalDate.parse("2019-12-01");
```
May be a bug here, but nothing serious :)

Chustinjeng reviewed (commented):
Nice work! Overall I felt your code was well-organised and easy to read, but I guess there are some careless mistakes but it happens 👍

leeianee committed:
Add assertions to Parser.java

There are several methods that require specific inputs and will fail if
these inputs are incorrect.

Using assertions will help to detect any mistakes in parsing inputs
before using them in these functions.

Let's document any assertions used in the future so that it is easy to
see what assertions have to be adhered to.

leeianee committed:
Improve code quality

There were cases of unnecessary code duplication and excessive nesting
in Parser.java.

Fixing these instances of poor code quality improves the code quality.

To maintain code quality, let's adhere to good code quality practices
such as reducing nesting and code duplication.

leeianee committed:
Merge pull request #1 from leeianee/branch-A-Assertions

Add assertions to Parser.java

leeianee committed:
Merge pull request #2 from leeianee/branch-A-CodeQuality

Improve code quality

leeianee committed:
Add Undo functionality

leeianee committed:
Add User Guide

leeianee committed:
Fix save file detection
